[Post-LocalSGD] Comm hook that runs allreduce for the first K steps

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52636 [Post-LocalSGD] Comm hook that runs allreduce for the first K steps**

This DDP communication hook runs allreduce to average gradients for the first K steps, and then stops any communication afterwards.

This hook is designed to support post-local SGD, by running gradient averaging before local SGD stage. It is user's responsibility to perform model parameter averaging at local SGD stage after K steps.

A new test `test_allreduce_first_k_steps_ddp_comm_hook_nccl` is added. Also refactors test_c10d.py by removing the `_test_allreduce_ddp_comm_hook_nccl`, which is already included by `_test_stateless_ddp_comm_hooks_nccl`, and renaming `test_default_ddp_comm_hooks_nccl`  as `test_stateless_ddp_comm_hooks_nccl`.

Differential Revision: [D26576654](https://our.internmc.facebook.com/intern/diff/D26576654/)

Commit:
[Post-LocalSGD] Comm hook that runs allreduce for the first K steps

This DDP communication hook runs allreduce to average gradients for the first K steps, and then stops any communication afterwards.

This hook is designed to support post-local SGD, by running gradient averaging before local SGD stage. It is user's responsibility to perform model parameter averaging at local SGD stage after K steps.

A new test `test_allreduce_first_k_steps_ddp_comm_hook_nccl` is added. Also refactors test_c10d.py by removing the `_test_allreduce_ddp_comm_hook_nccl`, which is already included by `_test_stateless_ddp_comm_hooks_nccl`, and renaming `test_default_ddp_comm_hooks_nccl`  as `test_stateless_ddp_comm_hooks_nccl`.

Lint is applied to test_c10d.py.

Differential Revision: [D26576654](https://our.internmc.facebook.com/intern/diff/D26576654/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1259,7 +1259,11 @@ void Reducer::finalize_backward() {
           "Expected bucket.future_work not to be null. "
           "This may indicate that communication hook was not properly installed.");
-      bucket.future_work->wait();
+      // Normally need to wait until the completion of communication hook,
+      // unless the communication hook skips allreduce and only does local accumulation for (post-)local SGD.
+      if (!bucket.future_work->completed()) {
+        bucket.future_work->wait();
+      }
```
I'm not sure why we need this change for post local sgd? Just calling `wait()` should work even for post local SGD since we're retuning a completed future?

SciPioneer replied:
We cannot call `wait()` again on a completed future returned by post local sgd. The training will be stuck.

SciPioneer replied:
Per offline discussion, will make `ivalue::Future::wait()` here conditional and avoid extra overheads due to an additional unnecessary lock.

SciPioneer replied:
Actually this change will fail some tests in `distributed_test.py`. With the help of @rohan-varma, realized that the returned type is `CUDAFuture`, which will be always a completed future and will still need to call `wait()` for synchronization.

Reverted this change.

pritamdamania87 commented on torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:
```diff
@@ -65,3 +67,78 @@ def decompress(fut):
+        state.maybe_increase_step(bucket)
+        return _allreduce_fut(group_to_use, input_tensors[0])
+    else:
+        # Afterwards K steps, directly return the input tensors as the output future.
+        ret_fut = torch.futures.Future()
+        ret_fut.set_result(input_tensors)
+        return ret_fut
```
Shouldn't we implement local SGD here where we now average gradients every H steps? I'm not sure why we leave this upto the user? It should be easy to implement that logic here right?

SciPioneer replied:
Local SGD needs to average parameters instead of gradients, and parameter averaging should be executed after `optimizer.step()`, so actually it cannot be implemented in the context of DDP at this time.

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -360,16 +361,23 @@ def test_compare_set(self):
         except Exception:
-            logging.error('Caught exception: \n{}exiting process with exit code: {}'
-                          .format(traceback.format_exc(), MultiProcessTestCase.TEST_ERROR_EXIT_CODE))
+            logging.error(
+                "Caught exception: \n{}exiting process with exit code: {}".format(
+                    traceback.format_exc(), MultiProcessTestCase.TEST_ERROR_EXIT_CODE
+                )
+            )
```
It's usually easier to review PRs if it doesn't contain formatting changes like these. Usually its better to send a separate PR with formatting changes.

SciPioneer replied:
Yeah, I somehow messed it up. Let me revert the irrelevant changes.

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -3822,9 +3825,36 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        # Get GPU model with the hook registered.
+        for use_error_feedback, warm_start in product([True, False], [True, False]):
```
What is the purpose of this for loop? Doesn't look like we're using `use_error_feedback` or `warm_start`?

SciPioneer replied:
Extra code copied from PowerSGD unit test. Removed now.

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -3822,9 +3825,36 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
+            )
+
+            # check whether the grads are equal to what DDP without hook would return.
+            self._run_and_verify_hook(gpu_model, 8, 0.25 * torch.ones(2, 2))
```
This only runs one iteration, so I'm not sure how this is testing an algorithm that probably needs multiple iterations?  Also, mathematically post local SGD wouldn't be equivalent to DDP without hooks in general right?

SciPioneer replied:
Currently `_run_and_verify_hook` only runs one step. To test it more thoroughly, need to provide another util function.

This test only tests one special case at this time, where post local SGD is equivalent to DDP if there are too few steps and local SGD does not start at all.

What about adding a TODO here to add another test later? I prefer exploring some local SGD algorithms first. It's still unclear how it will work with DDP.

Commit:
Update on "[Post-LocalSGD] Comm hook that runs allreduce for the first K steps"


This DDP communication hook runs allreduce to average gradients for the first K steps, and then stops any communication afterwards.

This hook is designed to support post-local SGD, by running gradient averaging before local SGD stage. It is user's responsibility to perform model parameter averaging at local SGD stage after K steps.

A new test `test_allreduce_first_k_steps_ddp_comm_hook_nccl` is added. Also refactors test_c10d.py by removing the `_test_allreduce_ddp_comm_hook_nccl`, which is already included by `_test_stateless_ddp_comm_hooks_nccl`, and renaming `test_default_ddp_comm_hooks_nccl`  as `test_stateless_ddp_comm_hooks_nccl`.

Lint is applied to test_c10d.py.

Differential Revision: [D26576654](https://our.internmc.facebook.com/intern/diff/D26576654/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:
```diff
@@ -65,3 +67,78 @@ def decompress(fut):
+            )
+        )
+
+    def maybe_increase_step(self, bucket):
```
nit: I remember we have a similar function for other hooks that keep track of iteration. Could we consolidate that logic, maybe by a simple counter function/class?

SciPioneer replied:
They are similar, applied to different state classes. The main difference is the log info and the variable name that represents the last iteration.

To consolidate them, need to expose a helper function as a new public API. However, this hook is very experimental, as it is still unclear to support post-local SGD with DDP, prefer no public API changes based on this hook at this time.

Added a TODO here.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:
```diff
@@ -65,3 +67,78 @@ def decompress(fut):
+            only exactly one tensor is stored in this bucket.
+
+    Returns:
+        Future handler of the communication, which updates the gradients in place.
```
nit: mention that this will just a completed future with input tensors (local grads) in the case step > k?

SciPioneer replied:
Good catch!

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -3819,6 +3799,36 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
+        state = default.AllReduceFirstKStepsState(
+            process_group=process_group,
+            # Should be equivalent to allreduce for the first K steps.
+            k=1,
```
I feel like we should test > 1 iterations and with a k > 1. For example, can we test n iterations and k = m, and for the last n - m steps, verify that the gradients are local-only?

SciPioneer replied:
Agree. Currently `_run_and_verify_hook` only runs one step. To test it more thoroughly, probably need to provide another util function.

Already added a TODO here to add more test cases later. I prefer exploring some local SGD algorithms first. It's still unclear how it will work with DDP. We can make this hook as a private method at this time.

pritamdamania87 replied:
I feel we should add the unit tests in the PR where we introduce the feature itself. You can always patch this PR locally and explore algorithms on top of it.

Or if we're only exploring this idea at this time and might not land this change, probably we can tag this PR as [WIP] or something indicating its not ready for review yet.

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -3866,18 +3876,18 @@ def _test_builtin_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
-    def test_builtin_ddp_comm_hooks_nccl(self):
-        self._test_builtin_ddp_comm_hooks_nccl()
+    def test_allreduce_first_k_steps_ddp_comm_hook_nccl(self):
+        self._test_allreduce_first_k_steps_ddp_comm_hook_nccl()
```
Should we test with gradient_as_bucket_view=True?

SciPioneer replied:
Added

zhaojuanmao commented:
DDP allows users to run some iteration with context manager "ddp_model.no_sync()", which should be simple enough to run allreduce for the first K steps. 
```
    for i in range(n): 
        if i < k: 
           context = model.no_sync()
       esle: 
           context = contextlib.nullcontext()
      with context: 
           training logic
```

I'm wondering why we need this hook? if users use this hook, it will increase memory usage as well as increase overhead of copy data btw future and bucket?

Commit:
Update on "[Post-LocalSGD] Comm hook that runs allreduce for the first K steps"


This DDP communication hook runs allreduce to average gradients for the first K steps, and then stops any communication afterwards.

This hook is designed to support post-local SGD, by running gradient averaging before local SGD stage. It is user's responsibility to perform model parameter averaging at local SGD stage after K steps.

A new test `test_allreduce_first_k_steps_ddp_comm_hook_nccl` is added. Also refactors test_c10d.py by removing the `_test_allreduce_ddp_comm_hook_nccl`, which is already included by `_test_stateless_ddp_comm_hooks_nccl`, and renaming `test_default_ddp_comm_hooks_nccl`  as `test_stateless_ddp_comm_hooks_nccl`.

Differential Revision: [D26576654](https://our.internmc.facebook.com/intern/diff/D26576654/)

[ghstack-poisoned]

Commit:
Update on "[Post-LocalSGD] Comm hook that runs allreduce for the first K steps"


This DDP communication hook runs allreduce to average gradients for the first K steps, and then stops any communication afterwards.

This hook is designed to support post-local SGD, by running gradient averaging before local SGD stage. It is user's responsibility to perform model parameter averaging at local SGD stage after K steps.

A new test `test_allreduce_first_k_steps_ddp_comm_hook_nccl` is added. Also refactors test_c10d.py by removing the `_test_allreduce_ddp_comm_hook_nccl`, which is already included by `_test_stateless_ddp_comm_hooks_nccl`, and renaming `test_default_ddp_comm_hooks_nccl`  as `test_stateless_ddp_comm_hooks_nccl`.

Differential Revision: [D26576654](https://our.internmc.facebook.com/intern/diff/D26576654/)

[ghstack-poisoned]

SciPioneer commented:
> DDP allows users to run some iteration with context manager "ddp_model.no_sync()", which should be simple enough to run allreduce for the first K steps.
> 
> ```
>     for i in range(n): 
>         if i < k: 
>            context = model.no_sync()
>        esle: 
>            context = contextlib.nullcontext()
>       with context: 
>            training logic
> ```
> 
> I'm wondering why we need this hook? if users use this hook, it will increase memory usage as well as increase overhead of copy data btw future and bucket?

The key difference is that if `no_sync` context manager is used, `optimizer.step()` must be separate out of the training loop. However local SGD needs to include it in the loop.

This hook does not clone any separate tensors. Since the underlying storage of the input tensors will be shared within the hook, I don't think it will increase memory usage.

SciPioneer commented:
To submit this PR, still need to write a unit test case where no more gradient averaging after K steps.

Close this PR since I find another better approach that does not need this communication hook.
